### PR TITLE
Support multi-gpu for ChatGLM-v2

### PR DESCRIPTION
**Pull request by sijunhe**

### CODE REVIEW

File: paddlenlp/transformers/chatglm_v2/modeling.py [hunk 1]
```diff
@@ -221,87 +215,90 @@ class SelfAttention(nn.Layer):
...
+        Assign kv heads to different GPUs in the Tensor Parallel Setup
+
+        Examples:
+            assign_kv_heads(num_kv_heads=1, num_gpus=2): [[0], [0]]
+            assign_kv_heads(num_kv_heads=2, num_gpus=2): [[0], [1]]
+            assign_kv_heads(num_kv_heads=4, num_gpus=2): [[0,1], [2,3]]
+            assign_kv_heads(num_kv_heads=1, num_gpus=4): [[0],[0],[0],[0]]
+            assign_kv_heads(num_kv_heads=2, num_gpus=4): [[0],[0],[1],[1]]
+            assign_kv_heads(num_kv_heads=4, num_gpus=4): [[0],[1],[2],[3]]
+        """
```
Comment (wj-Mcat) [inline paddlenlp/transformers/chatglm_v2/modeling.py]: 感觉这个函数可以作为function：因为self 从头到尾没用到，和模型无关。

File: paddlenlp/transformers/chatglm_v2/modeling.py [hunk 2]
```diff
@@ -318,21 +315,21 @@ def forward(self, hidden_states, attention_mask, rotary_pos_emb, kv_cache=None,
...
-                key_layer.shape[:2] + [self.num_attention_heads_per_partition, self.hidden_size_per_attention_head]
-            )
-            value_layer = value_layer.unsqueeze(-2)
-            value_layer = value_layer.tile(
-                [1, 1, 1, self.num_attention_heads_per_partition // self.num_multi_query_groups_per_partition, 1]
-            )
-            value_layer = value_layer.reshape(
-                value_layer.shape[:2] + [self.num_attention_heads_per_partition, self.hidden_size_per_attention_head]
-            )
+        if hasattr(self, "kv_indices"):
```
Comment (wj-Mcat) [inline paddlenlp/transformers/chatglm_v2/modeling.py]: hasattr 在动转静的过程中不太友好，会出现问题，建议换一种写法: `if self.config.tensor_parallel_degree > 1`来控制self.kv_indices。

File: paddlenlp/transformers/chatglm_v2/modeling.py [hunk 3]
```diff
@@ -744,21 +789,26 @@ def forward(
...
+            reshaped_labels = labels.reshape([-1])
+
+            if self.config.tensor_parallel_degree > 1 and self.config.tensor_parallel_output:
+                loss_fn = fleet.meta_parallel.ParallelCrossEntropy()
+            else:
+                loss_fn = nn.CrossEntropyLoss(reduction="none")
+
+            loss_mask = (labels != -100).astype("float32")
+            loss = loss_fn(reshaped_logits, reshaped_labels)
+            loss = paddle.sum(loss.reshape([-1]).cast(paddle.float32) * loss_mask.reshape([-1]).cast(paddle.float32))
```
Comment (wj-Mcat) [inline paddlenlp/transformers/chatglm_v2/modeling.py:813]: 我看原始代码是通过 shift-labels 计算 loss，这里直接默认不 shift，将这个过程提前至数据预处理部分。

我在想：

1. 如果 shift - labels 建议都前置，那是否要作为 paddlenlp 增量模型的标准，都建议这样来做，因为从精度和效率来讲都要高一点。
2. 如果没有这个标准，是否通过 flag 来兼容两者情况。

我比较倾向于1.

**Review by wj-Mcat — COMMENTED** [review]
我看 ChatGLM 中的单测非常少，建议在这里加一些generation 相关的单测，保证生成的稳定性。

File: paddlenlp/transformers/chatglm_v2/modeling.py (hunk 3 above)
Comment (sijunhe) [reply paddlenlp/transformers/chatglm_v2/modeling.py:813]: 我的本意也是1，直接新增的模型直接强制数据流中shift

**Review by wj-Mcat — APPROVED** [review]
LGTM
